src/include/sof/trace/trace.h

```diff
@@ -388,4 +369,24 @@ do {									\
 #define trace_error_comp(...) _trace_comp_build(trace_error_with_ids,	\
 						##__VA_ARGS__)
 
+#include <sof/spinlock.h>
```

Comment:
Includes cannot be that far in code. They should be first thing in file (the only exepction are `#if CONFIG`'s). Even if you have to do `#if some_config_for_traces` more than 1 time because of doing it, please do it, it's better to have includes grouped and just split with ifs.

Reply:
Mostly agreed although includes being that far down shouldn't actually have significant side effects. There is a precedent in either this file or interrupt.h (forgot which) with mid-file includes anyway so I wouldn't block but it's your choice I guess.

Reply:
@jajanusz OK, I'll look a bit more into it. The problem here seems to be that when spinlock debugging is on, there is a cyclic include dependency between `trace.h` and `spinlock.h`. Namely when `trace.h` gets included, it needs `spinlock.h` for resolving `spinlock_t`. But since debugging is enabled, spinlock.h will also include trace.h and would expect for those trace macros to exist. Moving that `#include` at the end of the file was the quick fix.

Reply:
Is this a situation of cyclic includes? The cyclic dependency needs to be broken down, I thought I fixed that already but apparently I only fixed it between spinlock and interrupt.

Guess another refactor on the spinlocks is needed. @jajanusz What do you think?

Reply:
@paulstelian97 That's right. This cyclic include dependency seems to have been introduced by commit 6e2cf6920bd5c62095e26e6751d3169a07c34b5c .

Reply:
@dragosht Indirect include? Because <sof/spinlock_t.h> is not <sof/spinlock.h> but the solution I had offered for the previous cyclic include dependency.

Reply:
@paulstelian97 That typo was later fixed by e465bca588328a801dd6954cbeba739d369ae3d5

Reply:
e465bca5 removed that separation, reintroducing the cyclic dependencies. @tlauda Must have not spotted that while reviewing, sorry.

Reply:
By the `far` I didn't mean number of lines, what is a real problem is introduction of macros/types/symbols before includes. If you really need them and there is some cycle to break just move common part to separate file.

Reply:
@paulstelian97 It is possible, we could miss that during review. I guess he wanted to refactor something and it worked ok without debug mode on and CI doesn't check that scenario. We really need CI to build more flavors of fw...

Reply:
Guess we can refer to #2318 in this sense. Anyway, now work has to be done to break the cyclic dependencies again.

Reply:
@paulstelian97 @jajanusz  Got it. I haven't checked that very thorough so I assumed that `spinlock_t.h` change was actually a typo not an accidental revert. @paulstelian97 Should I rework this patch, or would you like to revert the reverted code?

Reply:
I honestly don't know what would be better/easier. The separation is needed because the cyclic dependency is there so you should take a look in the reverted code. That header only included the declaration itself but not the other spinlock related functions, which were still in the full <spinlock.h>. So you should try to revert that part of said patch (not the full patch, the other changes do make sense).

Reply:
It would be better if trace wouldn't depend on spinlock routines at all. Why would it? It's bad design, but it's done for performance (there are inline functions in trace.h when debug is on). IMHO it's debug mode anyway and these inlines should be normal function that are compiled in some compilation unit (trace.c) and then trace.h will not have dependency on spinlock.h. That's how I'd do it.

Reply:
@dragosht If you can do that refactor that would be great, but if you don't have time, feel free to just file issue like '<some config> breaks compilation'

Reply:
@jajanusz If it's triggered during a verbose trace, performance is required even in debug mode, you don't want an Xrun storm caused by debug mode. So inlines may be required but what is the lock itself needed for?

Reply:
OK, good point, I didn't notice there is dependency even in non-debug build. Anyway these scenarios use trace only for reporting errors so they still can be moved out of inlined functions, because you optimize for good case. Exceptions are always done like that, no cost on usual scenario, but high cost when error/exception occur. Here cost will be 1 extra call and we will not have dependency on trace.h in spinlock.h.

Reply:
Alright, anyway I don't have enough time for the moment to tackle that (once I'm done with exams I'll really focus on the dynamic module loading thing)

Reply:
@jajanusz @paulstelian97 OK. I see there is quite some context to dig through for these changes. I'll just close this PR and raise an issue on this.